### DISCUSSION TITLE
ch5-web/ch5-03-middleware

### DISCUSSION THREAD

**Discussion by giscus** [chai2010/advanced-go-programming-book#631] [General]
# ch5-web/ch5-03-middleware

Go语言高级编程

https://chai2010.cn/advanced-go-programming-book/ch5-web/ch5-03-middleware.html

**Comment by wangfeng2500** [comment]
5.3.2 这段代码响应参数是错的
func timeMiddleware(next http.Handler) http.Handler {
    return http.HandlerFunc(func(wr http.ResponseWriter, r *http.Request) {
        timeStart := time.Now()

        // next handler
        next.ServeHTTP(wr, r)

        timeElapsed := time.Since(timeStart)
        logger.Println(timeElapsed)
    })
}
应该是func timeMiddleware(next http.Handler) http.HandlerFunc ，http.Handler是一个interface，需要实现serveHttp方法的

go 1.14

**Reply by wangfeng2500** [reply]
要么改为：
1. func timeMiddleware(next http.Handler) http.Handler
2. http.HandleFunc("/", timeMiddleware(http.HandlerFunc(hello)).ServeHTTP)

**Reply by typerfect** [reply]
你的理解是错误的，返回的是http.HandlerFunc类型，由于http.HandlerFunc类型实现了ServeHTTP方法，所以http.HandlerFunc可以认为是http.Handler。作者写的非常详细，你没有正确理解。
